### ISSUE TITLE
PyInstaller, issue to build a package with CUDA, CuDNN and Tensorflow on Windows 10

### ISSUE THREAD

**Issue by cberri** [pyinstaller/pyinstaller#7175]
Hi there,

First, thank you for developing and maintaining such a great tool. I am very new to pyinstaller and I apologies if I miss anything that might be obvious to you.

I am currently using pyinstaller to build a GUI to run a Deep Learning tool that uses CUDA, CuDNN and tensorflow. I am having my virtual environment in conda with the following dependency on Windows 10:

```yaml
1121 INFO: PyInstaller: 5.1
1122 INFO: Python: 3.8.13 (conda)
1129 INFO: Platform: Windows-10-10.0.19044-SP0

# To manually install the yml run: conda env create -f environment.yml
name: StarDistGUI
channels:
    - conda-forge
dependencies:
    - pip
    - python=3.8
    - cudatoolkit=11.3.1
    - cudnn=8.2
    - tensorflow=2.9.1=gpu_py38h4a6fda8_0
    - pyqt=5.15.7
    - pyqtwebengine=5.15.7
    - psutil
    - csvkit
    - setuptools
    - pyinstaller
    - hdf5
    - pip:
        - csbdeep==0.7.2
        - gputools==0.2.13
        - stardist==0.8.3
        - pyradiomics
        - connected-components-3d
        - ray
        - rais
```

I packaged my app with pyinstaller as below from PyCharm:

`pyinstaller --add-data="temp;./temp" --icon=stardist.ico stardistGUI.py  --hidden-import=skimage.filters.edges --collect-all tensorflow --noconfirm --clean`

`temp` contains my .ui file

Everything seems to work and I get the *build* and *dist* directories. I can start my tool from the *dist* directory without any problem.

Here comes the issue:

I have locally installed on my Windows 10 machine CUDA 11.0 with CuDNN 8.1.x and CUDA 11.3 with CuDNN 8.2.x. It should not matter because I provide CUDA and CuDNN for the tool in the conda environment:

```
- cudatoolkit=11.3.1
- cudnn=8.2
```

I build the tool with pyinstaller and things get strange. If I have set my windows 10 env variables for CUDA V11.0 with CuDNN 8.1.x the build tool pops the following error message:

```
2022-10-19 16:23:52.416036: E tensorflow/stream_executor/cuda/cuda_dnn.cc:377] Loaded runtime CuDNN library: 8.1.0 but source was compiled with: 8.2.1.  CuDNN library needs to have matching major version and equal or higher minor version. If using a binary install, upgrade your CuDNN library.  If building from sources, make sure the library loaded at runtime is compatible with the version specified during compile configuration.
2022-10-19 16:23:52.425421: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at conv_ops.cc:1120 : UNIMPLEMENTED: DNN library is not found.
```

I printed the following after building the package with pyinstaller: GPU is recognized, CUDA seems all right (64_113), CuDNN I am honestly not sure (64_8). In conda I have CuDNN 8.2.x but locally 8.1.x.

```python
from tensorflow.python.client import device_lib
print(device_lib.list_local_devices())

2022-10-19 16:08:35.405699: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /device:GPU:0 with 7437 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 3080, pci bus id: 0000:68:00.0, compute capability: 8.6
[name: "/device:CPU:0"
device_type: "CPU"
memory_limit: 268435456
locality {
}
incarnation: 16001663011293130701
xla_global_id: -1
, name: "/device:GPU:0"
device_type: "GPU"
memory_limit: 7799234560
locality {
  bus_id: 1
  links {
  }
}
incarnation: 4786325627852168319
physical_device_desc: "device: 0, name: NVIDIA GeForce RTX 3080, pci bus id: 0000:68:00.0, compute capability: 8.6"
xla_global_id: 416903419
]

import tensorflow as tf
build = tf.sysconfig.get_build_info()
print(build['cuda_version'])
print(build['cudnn_version'])

64_113
64_8
```

If I change my local env to CUDA 11.3 (set Windows 10 env variables to V11.3) everything works fine after building the app with pyinstaller. No error message it runs properly with the right version of CuDNN (8.2.x).

I think I am missing something or at least pyinstaller builds my package with the wrong CuDNN libraries or somehow looks only at my local installation and not what is in my conda environment. From PyCharm I can run successfully the tool independently from what I have in my local environment. After a week I have no more ideas how to solve the issue and I would really appreciate your suggestions.

If I miss any meaningful information for you please let me know and also if something is not clear in my question. I tried my best to explain the problem but it is not really straightforward.

Thank you in advance for your help.

Best regards,

Carlo

**Comment by rokm** [comment]
To check which CUDA libs are collected (and from where), try opening `build/<name>/Analysis-00.toc` and find corresponding DLL entries. 

You mention "changing Windows 10 env variables" to switch between CUDA versions - I assume you are modifying `PATH`? Are you prepending or appending? The order of DLL resolution depends on the order in `PATH`, which may be why CUDA DLLs from local installation are picked over the conda ones.

**Comment by rokm** [comment]
But since `tensorflow` loads CUDA shared libraries dynamically, it could just as well be that PyInstaller does not collect them (since they cannot be inferred via binary dependency analysis), and ends up using whatever you have set up in the environment. Have you checked that your frozen application in `dist` in fact contains CUDA and CuDNN DLLs?

**Comment by cberri** [comment]
Looking at the .toc file everything looks good to me. It looks at the right conda env. I can also attached the complete file in case.

```
...

'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\__init__.py',
   'PYMODULE'),
  ('numba.cuda.compiler',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\compiler.py',
   'PYMODULE'),
  ('numba.cuda.descriptor',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\descriptor.py',
   'PYMODULE'),
  ('numba.cuda.target',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\target.py',
   'PYMODULE'),
  ('numba.cuda.dispatcher',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\dispatcher.py',
   'PYMODULE'),
  ('numba.cuda.types',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\types.py',
   'PYMODULE'),
  ('numba.cuda.errors',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\errors.py',
   'PYMODULE'),
  ('numba.cuda.args',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\args.py',
   'PYMODULE'),
  ('numba.cuda.api',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\api.py',
   'PYMODULE'),
  ('numba.cuda.device_init',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\device_init.py',
   'PYMODULE'),
  ('numba.cuda.intrinsic_wrapper',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\intrinsic_wrapper.py',
   'PYMODULE'),
  ('numba.cuda.decorators',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\decorators.py',
   'PYMODULE'),
  ('numba.cuda.simulator.kernel',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\kernel.py',
   'PYMODULE'),
  ('numba.cuda.simulator',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\__init__.py',
   'PYMODULE'),
  ('numba.cuda.simulator.compiler',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\compiler.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv.runtime',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\runtime.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv.devices',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\devices.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\__init__.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv.nvvm',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\nvvm.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv.error',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\error.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv.drvapi',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\drvapi.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv.driver',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\driver.py',
   'PYMODULE'),
  ('numba.cuda.simulator.reduction',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\reduction.py',
   'PYMODULE'),
  ('numba.cuda.simulator.vector_types',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\vector_types.py',
   'PYMODULE'),
  ('numba.cuda.simulator.api',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\api.py',
   'PYMODULE'),
  ('numba.cuda.simulator.kernelapi',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\kernelapi.py',
   'PYMODULE'),
  ('numba.cuda.simulator.cudadrv.devicearray',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator\\cudadrv\\devicearray.py',
   'PYMODULE'),
  ('numba.cuda.stubs',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\stubs.py',
   'PYMODULE'),
  ('numba.cuda.initialize',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\initialize.py',
   'PYMODULE'),
  ('numba.cuda.models',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\models.py',
   'PYMODULE'),
  ('numba.cuda.mathimpl',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\mathimpl.py',
   'PYMODULE'),
  ('numba.cuda.libdeviceimpl',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\libdeviceimpl.py',
   'PYMODULE'),
  ('numba.cuda.printimpl',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\printimpl.py',
   'PYMODULE'),
  ('numba.cuda.cudaimpl',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\cudaimpl.py',
   'PYMODULE'),
  ('numba.cuda.vector_types',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\vector_types.py',
   'PYMODULE'),
  ('numba.cuda.libdevicedecl',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\libdevicedecl.py',
   'PYMODULE'),
  ('numba.cuda.libdevicefuncs',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\libdevicefuncs.py',
   'PYMODULE'),
  ('numba.cuda.libdevice',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\libdevice.py',
   'PYMODULE'),
  ('numba.cuda.cudamath',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\cudamath.py',
   'PYMODULE'),
  ('numba.cuda.cudadecl',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\cudadecl.py',
   'PYMODULE'),
  ('numba.cuda.nvvmutils',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\nvvmutils.py',
   'PYMODULE'),
  ('numba.cuda.codegen',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\codegen.py',
   'PYMODULE'),
  ('numba.cuda.simulator_init',
   'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator_init.py',
   'PYMODULE'),

...
```


With changing the env variables in Windows I mean I replace the version of cuda in the existing variables path.

C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0
to 
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3


In the dist directory I can find CUDA and CuDNN. Here is the path:

`\dist\stardistGUI\tensorflow\include\tensorflow\stream_executor\cuda`

I guess should be the right location to look at.

**Comment by rokm** [comment]
> ('numba.cuda.simulator_init',
>    'C:\\Users\\CarloBeretta\\anaconda3\\envs\\StarDistGUI\\lib\\site-packages\\numba\\cuda\\simulator_init.py',
>    'PYMODULE'),

These are python modules, not DLLs.

> `\dist\stardistGUI\tensorflow\include\tensorflow\stream_executor\cuda`

This either contains python modules or extensions, but not CUDA DLLs. EDIT: it actually contains C/C++ headers.

---

Take a look into your `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0` or ` C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3` for the DLL files that comprise a CUDA environment. If they are collected, they would likely be so into the top-level application directory. 

If they are not collected (they likely are not due to dynamic loading), your application will end up depending on CUDA being installed (and available via `PATH`) on the target system.

**Comment by cberri** [comment]
I see and I am sorry for the confusion. I looked at both:

`C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\bin`

`C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3\bin`

In the main `dist/<name>/` directory I can not find any DLL file CUDA related. Should I simply copy the file from my  `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3\bin` to the dist directory?

Not sure if it makes any sense or there is a better way.

**Comment by rokm** [comment]
> Should I simply copy the file from my C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3\bin to the dist directory?

Yeah, while you're trying to figure out how to get it working, you can manually copy the DLLs over. Ideally, you should also remove the CUDA from PATH before you try to run your application, so you know that you are not missing any of the DLLs. Once you know which ones you need, you can create a list in the .spec file and pass it to `binaries` or `datas`; and then switch to building via the .spec file instead of .py file.

In my experience, though, CUDA deployments can be tricky, especially if kernels need to be re-compiled; in which case the ptxas compiler executable also needs to be accessible. So it may be easier to require that the target environment provides its own CUDA installation, and just copy the CuDNN DLL into your dist (provided the CuDNN license allows redistribution of the DLL).

**Comment by cberri** [comment]
THANK YOU. Let me try that on a fresh system without CUDA installed. I will post back the results.

**Comment by cberri** [comment]
Ok, I could not wait. I copied the DLL into the  `dist/<name>/` directory. The error is gone. I will give a try on a clean system just to be sure I copied everything that it is necessary and post the results back. Really THANK YOU for your help.

I didn't included in my copy and paste the ptxas.exe file but I guess I will find an issue if it is not included in the dist directory on a clean system.

**Comment by cberri** [comment]
Dear @rokm, a quick update. On Windows 10 without CUDA installed everything seems to work pretty well. I included in my "copy and paste" the  ptxas.exe file as well. It solved some warnings.

My current command to build the package is:

```
pyinstaller --add-data="temp;./temp" --add-data="dll/*;." --add-data="icon;./icon" --icon=stardist.ico stardistGUI.py --hidden-import=skimage.filters.edges --hidden import=msgpack --onefile --noconfirm --clean
```

`--add-data="dll/*;."` the directory with all the dll file missing.

I built the application with `--onefile` or as a directory without major differences.

I am still having trouble with other parts but as far as I can see your suggestion solved the issues for CUDA.

Thank you again.

Carlo
